File: n_queens_backtrack.py
class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        result = []
        board = [[0 for _ in range(n)]for _ in range(n)]
        if not n:
            return result
        self.backtrack(n, 0, result, board)
        return result
    
    def backtrack(self, n, row, result, board):
        def isSafe(row, col):
            #column up
            for i in range(0,row):
                if board[i][col] == 1:
                    return False
            #diagonal up left
            i, j = row-1, col-1
            while(i>=0 and j>=0):
                if board[i][j] == 1:
                    return False
                i-=1
                j-=1
            #diagonal up right
            i, j = row-1, col+1
            while(i>=0 and j<n):
                if board[i][j] == 1:
                    return False
                i-=1
                j+=1
            return True
        #base
        if row == n:
            temp = []
            for i in range(n):
                string = ""
                for j in range(n):
                    if board[i][j] == 1:
                        string+='Q'
                    else:
                        string+='.'
                temp.append(string)
            result.append(temp)
            return
        #logic
        for i in range(n):
            if isSafe(row, i):
                #action
                board[row][i] = 1
                #recurse
                self.backtrack(n, row+1, result, board)
                #backtrack
                board[row][i] = 0


#time complexity - O(n!) as every cell of the board is tested out

#space complexity - O(n)

#all test cases passed

File: NQueens_BackTrack.java
import java.util.ArrayList;
import java.util.List;

/*
## Problem1 
N Queens(https://leetcode.com/problems/n-queens/)

Time Complexity :   O (N!) 
Space Complexity :  O (N^2) 
Did this code successfully run on Leetcode :    Yes (51. N-Queens)
Any problem you faced while coding this :       No

 */
// Input: n = 4
// Output: [[".Q..","...Q","Q...","..Q."],["..Q.","Q...","...Q",".Q.."]]

class NQueens_BackTrack {
    List<List<String>> result;
    int m;
    
    public List<List<String>> solveNQueens(int n) {
        result = new ArrayList<>();
        m=n;
        
        boolean[][] board = new boolean[n][n];
        
        // null case
        if(n == 0)
            return result;
        
        backtrack(board, 0);
        return result;
    }
    private void backtrack(boolean[][] board, int r){
        // base
        if( r == m){
            List<String> li = new ArrayList<>();
            
            for(int i=0; i< m; i++){
                StringBuilder sb = new StringBuilder();
                for(int j=0; j<m; j++){
                    if(board[i][j]){
                        sb.append("Q");
                    }else{
                        sb.append(".");
                    }
                }
                li.add(sb.toString());
            }
            result.add(li); 
        }
        
        //logic
        for(int c = 0; c< m; c++){
            if(isSafe(board, r, c)){
                // action
                board[r][c] = true;
                
                // recurse
                backtrack(board, r+1);
                
                // backtrack
                board[r][c] = false;
            }
        }
    }
    
    private boolean isSafe(boolean[][] board, int r, int c){
        // check if same column
        for(int i=0; i< r; i++){
            if(board[i][c])
                return false;
        }
        
        // check upper left diagonal
        int i = r;
        int j = c;
        while( i >= 0 && j>= 0){
            if(board[i][j])
                return false;
            i--;
            j--;
        }
        
        // check upper right diagonal
        i = r;
        j = c;
        while( i >= 0 && j < m){
            if(board[i][j])
                return false;
            i--;
            j++;
        }
        
        return true;
    }
}